### ISSUE TITLE
Unable to connect to localhost

### ISSUE THREAD

**Issue by andrewryno** [bobthecow/genghis#109]
Noticed this after upgrading from one of the 2.2 versions to 2.3.4. Doesn't matter what kind of DSN I try, it won't connect.

Mongo v2.2.3

Log after startup:

```
Wed May  1 14:53:54 [initandlisten] MongoDB starting : pid=7652 port=27017 dbpath=/usr/local/var/mongodb 64-bit host=Ryno.local
           14:53:54 [initandlisten] 
           14:53:54 [initandlisten] ** WARNING: soft rlimits too low. Number of files is 256, should be at least 1000
           14:53:54 [initandlisten] db version v2.2.3, pdfile version 4.5
           14:53:54 [initandlisten] git version: f570771a5d8a3846eb7586eaffcf4c2f4a96bf08
           14:53:54 [initandlisten] build info: Darwin bs-osx-106-x86-64-1.local 10.8.0 Darwin Kernel Version 10.8.0: Tue Jun  7 16:33:36 PDT 2011; root:xnu-1504.15.3~1/RELEASE_I386 i386 BOOST_LIB_VERSION=1_49
           14:53:54 [initandlisten] options: { bind_ip: "127.0.0.1", config: "/usr/local/etc/mongod.conf", dbpath: "/usr/local/var/mongodb", logappend: "true", logpath: "/usr/local/var/log/mongodb/mongo.log", rest: true }
           14:53:54 [initandlisten] journal dir=/usr/local/var/mongodb/journal
           14:53:54 [initandlisten] recover : no journal files present, no recovery needed
           14:53:54 [websvr] admin web console waiting for connections on port 28017
           14:53:54 [initandlisten] waiting for connections on port 27017
```

I've tried localhost, localhost:27017, 127.0.0.1, 127.0.0.1:27017, etc. Can't figure out what's going wrong. It seems to be writing using the PHP Mongo extension, so it's definitely available (my DSN there is `mongodb://127.0.0.1:27017`).

**Comment by bobthecow** [comment]
This is a strange one. You're using the PHP backend for Genghis, and it won't connect with the same DSN that you're using in straight PHP code?

**Comment by andrewryno** [comment]
Yup, I've copied it directly. DSN in PHP is `mongodb://127.0.0.1:27017`, Genghis adds a server called `127.0.0.1`, so it parsed it correctly. Then it tells me it's unable to connect.

I've tried using it in my LAMP setup locally, as well as using the PHP 5.4 built-in server directly.

Wondering if my Vagrant setup was causing this (which is no longer setup), so ran a scan for ports... looks like this should be fine (mongod is listening on the right port).

```
$ lsof -i -P | grep 27017
httpd      3129 andrew   38u  IPv4 0xcb68fa704acbc11b      0t0  TCP localhost:50023->localhost:27017 (CLOSE_WAIT)
httpd      3462 andrew   38u  IPv4 0xcb68fa704acb8e93      0t0  TCP localhost:50025->localhost:27017 (CLOSE_WAIT)
mongod     9471 andrew    6u  IPv4 0xcb68fa704c2edb73      0t0  TCP localhost:27017 (LISTEN)
```

**Comment by andrewryno** [comment]
Ok, so while my Mongo code doesn't seem to be throwing any exceptions, the `connected` class variable is returning false. So this appears to be a problem with my setup, not Genghis.

```
$ php -a
php > $client = new MongoClient('mongodb://localhost:27017');
php > var_dump($client);
class MongoClient#4 (4) {
  public $connected =>
  bool(false)
  public $status =>
  NULL
  protected $server =>
  NULL
  protected $persistent =>
  NULL
}
```

_sigh_. Closing this, and I'll figure out my problem separately. Thanks!

**Comment by andrewryno** [comment]
Just in case someone has a similar problem. Basically I updated PHP and mongo via homebrew, uninstalled the mongo PHP driver using pecl and installed it through homebrew-php. Updated all my paths and it started working. I'm really not sure what the problem was but I think some of my versions were out of sync or not installed correctly...

**Comment by bobthecow** [comment]
Great. Glad to hear you got it resolved. Now I just need to fix that other one for you.

(In the interim, clear your cookies and it'll go away).

**Comment by michaldudek** [comment]
Just had exactly the same issue and `brew install php53-mongo` solved it. Thank you!

**Comment by nicktacular** [comment]
I just experienced the same issue. All I did was upgrade from 2.2.0 to 2.3.5 and suddenly all my connections resulted in errors. I was connected to MongoDB 2.4.4 and 2.2.3 on separate servers. My PHP version 5.3.19, PHP mongo is 1.3.1 and I'm able to connect fine using the Mongo driver in PHP just fine:

``` php
<?php

$servers = array('mongodb://localhost:27017', 'mongodb://localhost:20202');

foreach ($servers as $server)
{
        $m = new MongoClient($server);
        echo "Connected to: $server\n";
        $sver = $m->local->execute("db.serverStatus().version");
        $sver = $sver['retval'];
        echo "Server version: $sver\n";
        echo "PHP Mongo Client Version: " . $m::VERSION . PHP_EOL;
        echo ($m->connected?'CONNECTED!':'Not connected :(') . PHP_EOL;
        $dbs = $m->listDBs();
        echo "Size: {$dbs['totalSize']}\n";
        echo ($dbs['ok']?"OK!\n":"Not ok :(\n");
        echo "DBs: " . count($dbs['databases']) . "\n\n";
        $m->close(true);
}
```

The output is:

```
$ php mongotest.php

Connected to: mongodb://localhost:27017
Server version: 2.4.4-pre-
PHP Mongo Client Version: 1.3.1
CONNECTED!
Size: 2315255808
OK!
DBs: 10

Connected to: mongodb://localhost:20202
Server version: 2.2.3
PHP Mongo Client Version: 1.3.1
CONNECTED!
Size: 1023410176
OK!
DBs: 2
```

However, in Genghis 2.3.5 all of these result in "Unable to connect to Mongo server at (host)":
- localhost
- localhost:20202
- 127.0.0.1
- 127.0.0.1:20202

All I did was update Genghis from 2.2.0 to 2.3.5. Nothing else changed and all my Mongo apps continue to work correctly. Any ideas here?

**Comment by nicktacular** [comment]
Just upgraded to the PHP Mongo Client v1.3.7. Output from my script is still successful:

```
Connected to: mongodb://localhost:27017
Server version: 2.4.4-pre-
PHP Mongo Client Version: 1.3.7
CONNECTED!
Size: 2315255808
OK!
DBs: 10

Connected to: mongodb://localhost:20202
Server version: 2.2.3
PHP Mongo Client Version: 1.3.7
CONNECTED!
Size: 1023410176
OK!
DBs: 2
```

Genghis still not connecting to any of my servers.

**Comment by nicktacular** [comment]
What?! And the 5th attempt at an httpd restart solves the problem...???

Why am I a programmer again?

:computer: :gun:

**Comment by bobthecow** [comment]
@nicktacular Yeah. I hate computers :(

Had you tried restarting apache before? Any idea what it might have been?

**Comment by nicktacular** [comment]
Not really sure. The only thought that gives me calm is that while I was trying to restart apache using Virtual Host X, it wasn't actually doing anything. Maybe a s/w issue w/ Virtual Host X that compounded the confusion. Otherwise, I give up :smile:

**Comment by alexlavr** [comment]
Just ran yum update and got newer mongodb and apache among other things. So I got the same problem that I could connect fine myself from PHP but not using Genghis. There is a problem with the timeout option when connecting to mongodb. So I changed  

``` php
$this->connection = new Mongo($this->dsn, array_merge(array('connectTimeoutMS' => 1000), $this->options));
```

 to 

``` php
 $this->connection = new Mongo($this->dsn, $this->options);
```

And not it works but I'm still looking for a better solution. 

The source file responsible for this is here:
https://github.com/bobthecow/genghis/blob/master/src/php/Genghis/Models/Server.php#L63
